### PR TITLE
Add support of numpy 2.0.0b1

### PR DESCRIPTION
**Pull request by duburcqa**
Fixes https://github.com/stack-of-tasks/eigenpy/issues/447

It requires a follow-up PR in Boost::Python itself [here](https://github.com/boostorg/python/pull/432), but it does not prevent from merging this PR.

### CODE REVIEW

**Review by jcarpent — CHANGES_REQUESTED** [review]
Could you avoid renaming variables and perform the fix?

File: src/register.cpp [hunk 1]
```diff
@@ -52,25 +52,25 @@ int Register::registerNewType(
...
       NPY_NEEDS_PYAPI | NPY_USE_GETITEM | NPY_USE_SETITEM | NPY_NEEDS_INIT;
-  descr.type_num = 0;
-  descr.names = 0;
-  descr.fields = 0;
-  descr.alignment =
-      alignement;  // call_PyArray_DescrFromType(NPY_OBJECT)->alignment;
+  proto.type_num = 0;
+  proto.names = 0;
+  proto.fields = 0;
+  proto.alignment =
```
Comment (jcarpent) [inline src/register.cpp:68]: Changing to `proto` does not improve the readability.

Comment (jcarpent) [reply src/register.cpp:68]: Please revert it.

**Comment by jcarpent** [comment]
@duburcqa Thanks for the contribution to support the latest NumPy change. Yet, it seems important to first extend the support of EigenPy, then change some variable names if they improve the readability. Please make these two stages into two different PRs.

File: src/register.cpp (hunk 1 above)
Comment (duburcqa) [reply src/register.cpp:68]: ok. I was following numpy naming convention for proto-description of arrays.

File: src/register.cpp [hunk 2]
```diff
@@ -82,22 +82,22 @@ int Register::registerNewType(
...
   instance().type_to_py_type_bindings.insert(
       std::make_pair(type_info_ptr, py_type_ptr));
-  instance().py_array_descr_bindings[py_type_ptr] = new_descr;
-  instance().py_array_code_bindings[py_type_ptr] = code;
+  instance().py_array_descr_bindings[py_type_ptr] = descr;
+  instance().py_array_code_bindings[py_type_ptr] = type_num;
 
   //      PyArray_RegisterCanCast(descr,NPY_OBJECT,NPY_NOSCALAR);
-  return code;
+  return type_num;
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply src/register.cpp:100]: yes, it is just wrong to consider it is a return code.

**Comment by duburcqa** [comment]
I will fix support of `Python < 3.9` due to [unsupported Py_SET_TYPE](https://peps.python.org/pep-0674/).

**Comment by duburcqa** [comment]
The CI is passing !

**Review by jcarpent — APPROVED** [review]
Thanks a lot @duburcqa. Very nice push forward !

**Comment by jakirkham** [comment]
Thank you both for your hard work here! 🙏

Am curious, when building wheels for 3.5.0 or 3.5.1, were they built with NumPy 1 or 2?

**Comment by duburcqa** [comment]
Numpy 1, there is still work to do to fully support numpy 2 in pre-compiled binaries.

**Comment by jakirkham** [comment]
Ok thanks Alexis! 🙏

**Pull request merged**